Backports v0.21.0 (3) (#40977)
LLNL/spack: CHANGELOG.md

# v0.21.0 (2023-11-10)

`v0.21.0` is a major feature release.

## Features in this release

> 1. **Better error messages with Condition chaining**

Author: upper case intended?

Reviewer: ```suggestion
1. **Better error messages with condition chaining**
```

In v0.18, we added better error messages that could tell you what problem happened, but
they couldn't tell you *why* it happened. `0.21` adds *condition chaining* to the
solver, and Spack can now trace back through the conditions that led to an error and
build a tree of causes potential causes and where they came from.  For example:

```console
$ spack solve hdf5 ^cmake@3.0.1
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'cmake@3.0.1'
   2. Cannot satisfy 'cmake@3.0.1'
        required because hdf5 ^cmake@3.0.1 requested from CLI
   3. Cannot satisfy 'cmake@3.18:' and 'cmake@3.0.1
        required because hdf5 ^cmake@3.0.1 requested from CLI
        required because hdf5 depends on cmake@3.18: when @1.13:
          required because hdf5 ^cmake@3.0.1 requested from CLI
   4. Cannot satisfy 'cmake@3.12:' and 'cmake@3.0.1
        required because hdf5 depends on cmake@3.12:
          required because hdf5 ^cmake@3.0.1 requested from CLI
        required because hdf5 ^cmake@3.0.1 requested from CLI
```

More details in #40173.

> 2. **OCI buildcaches**

Author: build cache or buildcache?

Reviewer: ```suggestion
2. **OCI build caches**
```

You can now use an arbitrary [OCI](https://opencontainers.org) registry as a build
cache:

```console
$ spack mirror add my_registry oci://user/image # Dockerhub
$ spack mirror add my_registry oci://ghcr.io/haampie/spack-test # GHCR
$ spack mirror set --push --oci-username ... --oci-password ... my_registry  # set login credentials
$ spack buildcache push my_registry [specs...]
```

> And you can add optinally add a base image to get *runnable* images:

Author: ```suggestion
And you can add optionally add a base image to get *runnable* images:
```

```python
    license("Apache-2.0")
```

Licenses use [SPDX identifiers](https://spdx.org/licenses), and you can use SPDX
expressions to combine them:

```python
    license("Apache-2.0 OR MIT")
```

Like other directives in Spack, it's conditional, so you can handle complex cases like
Spack itself:

```python
   license("LGPL-2.1", when="@:0.11")
   license("Apache-2.0 OR MIT", when="@0.12:")
```

More details in #39346, #40598.

6. **`spack deconcretize` command**

We are getting close to having a `spack update` command for environments, but we're not
quite there yet. This is the next best thing. `spack deconcretize` gives you control
over what you want to update in an already concrete environment. If you have an
environment built with, say, `meson`, and you want to update your `meson` version, you
can run:

```console
spack deconcretize meson
```

and have everything that depends on `meson` rebuilt the next time you run `spack
concretize`. In a future Spack version, we'll handle all of this in a single command,
but for now you can use this to drop bits of your lockfile and resolve your dependencies
again. More in #38803.

7. **New `spack info` UI**

> The venerable `spack info` commannd was looking shabby compared to the rest of Spack's

Author: ```suggestion
The venerable `spack info` command was looking shabby compared to the rest of Spack's
```

        required because hdf5 ^cmake@3.0.1 requested from CLI
        required because hdf5 depends on cmake@3.18: when @1.13:
          required because hdf5 ^cmake@3.0.1 requested from CLI
   4. Cannot satisfy 'cmake@3.12:' and 'cmake@3.0.1
        required because hdf5 depends on cmake@3.12:
          required because hdf5 ^cmake@3.0.1 requested from CLI
        required because hdf5 ^cmake@3.0.1 requested from CLI
```

More details in #40173.

2. **OCI buildcaches**

You can now use an arbitrary [OCI](https://opencontainers.org) registry as a build
cache:

```console
$ spack mirror add my_registry oci://user/image # Dockerhub
$ spack mirror add my_registry oci://ghcr.io/haampie/spack-test # GHCR
$ spack mirror set --push --oci-username ... --oci-password ... my_registry  # set login credentials
$ spack buildcache push my_registry [specs...]
```

And you can add optinally add a base image to get *runnable* images:

```console
$ spack buildcache push --base-image ubuntu:23.04 my_registry python
Pushed ... as [image]:python-3.11.2-65txfcpqbmpawclvtasuog4yzmxwaoia.spack

$ docker run --rm -it [image]:python-3.11.2-65txfcpqbmpawclvtasuog4yzmxwaoia.spack
```

> More details in #38358.

Author: ```suggestion
This creates a container image from the Spack installations on the host system, without the
need to run `spack install` from a `Dockerfile` or `sif` file. It also addresses the inconvenience
of losing binaries of dependencies when `RUN spack install` fails inside `docker build`.

Further, the container image layers and build cache tarballs are the same files. This means that
`spack install` and `docker pull` use the exact same underlying binaries. If you previously used
`spack install` inside of `docker build`, this feature helps you save storage by a factor two.

More details in #38358.
```

  not have to create (#40756).
* `spack find -H` / `--hashes`: a new shortcut for piping `spack find` output to
  other commands (#38663)
* Add `spack checksum --verify`, fix `--add` (#38458)
* New `default_args` context manager factors out common args for directives (#39964)
* `spack compiler find --[no]-mixed-toolchain` lets you easily mix `clang` and
  `gfortran` on Linux (#40902)

## Performance improvements

* `spack external find` execution is now much faster (#39843)
* `spack location -i` now much faster on success (#40898)
* Drop redundant rpaths post install (#38976)
* ASP-based solver: avoid cycles in clingo using hidden directive (#40720)
* Fix multiple quadratic complexity issues in environments (#38771)

## Other new features of note

* archspec: update to v0.2.2, support for Sapphire Rapids, Power10, Neoverse V2 (#40917)
* Propagate variants across nodes that don't have that variant (#38512)
* Implement fish completion (#29549)
* Can now distinguish between source/binary mirror; don't ping mirror.spack.io as much (#34523)
* Improve status reporting on install (add [n/total] display) (#37903)

## Windows

> This release has the best windows support of any Spack release yet, with numerous

Author: ```suggestion
This release has the best Windows support of any Spack release yet, with numerous
```

improvements and much larger swaths of tests passing:

* MSVC and SDK improvements (#37711, #37930, #38500, #39823, #39180)
* Windows external finding: update default paths; treat .bat as executable on Windows (#39850)
* Windows decompression: fix removal of intermediate file (#38958)
* Windows: executable/path handling (#37762)
* Windows build systems: use ninja and enable tests (#33589)
* Windows testing (#36970, #36972, #36973, #36840, #36977, #36792, #36834, #34696, #36971)
* Windows PowerShell support (#39118, #37951)
* Windows symlinking and libraries (#39933, #38599, #34701, #38578, #34701)

## Notable refactors
* User-specified flags take precedence over others in Spack compiler wrappers (#37376)
* Improve setup of build, run, and test environments (#35737, #40916)

> * AutotoolsPackage / MakefilePackage: add gmake build dependency (#40380)

Author: ```suggestion
* `make` is no longer a required system dependency of Spack (#40380)
```

* Support Python 3.12 (#40404, #40155, #40153)
* docs: Replace package list with packages.spack.io (#40251)
* Drop Python 2 constructs in Spack (#38720, #38718, #38703)

> * Remove `LazyReference` from code (#38944)

Author: ```suggestion
```
dunno if that's super relevant?

## Binary cache and stack updates
* e4s arm stack: duplicate and target neoverse v1 (#40369)
* Add macOS ML CI stacks (#36586)
* E4S Cray CI Stack (#37837)
* e4s cray: expand spec list (#38947)
* e4s cray sles ci: expand spec list (#39081)

## Removals, deprecations, and syntax changes
* ASP: targets, compilers and providers soft-preferences are only global (#31261)
* Parser: fix ambiguity with whitespace in version ranges (#40344)
* Module file generation is disabled by default; you'll need to enable it to use it (#37258)
* Remove deprecated "extra_instructions" option for containers (#40365)
* Stand-alone test feature deprecation postponed to v0.22 (#40600)
* buildcache push: make `--allow-root` the default and deprecate the option (#38878)

## Notable Bugfixes
* Bugfix: propagation of multivalued variants (#39833)
* Allow `/` in git versions (#39398)
* Fetch & patch: actually acquire stage lock, and many more issues (#38903)

> * Environment/depfile: fix bug with Git hash versions (attempt #2) (#37560)

Author: ```suggestion
* Environment/depfile: better escaping of targets with Git versions (#37560)
```

```console
$ spack solve hdf5 ^cmake@3.0.1
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'cmake@3.0.1'
   2. Cannot satisfy 'cmake@3.0.1'
        required because hdf5 ^cmake@3.0.1 requested from CLI
   3. Cannot satisfy 'cmake@3.18:' and 'cmake@3.0.1
        required because hdf5 ^cmake@3.0.1 requested from CLI
        required because hdf5 depends on cmake@3.18: when @1.13:
          required because hdf5 ^cmake@3.0.1 requested from CLI
   4. Cannot satisfy 'cmake@3.12:' and 'cmake@3.0.1
        required because hdf5 depends on cmake@3.12:
          required because hdf5 ^cmake@3.0.1 requested from CLI
        required because hdf5 ^cmake@3.0.1 requested from CLI
```

More details in #40173.

2. **OCI build caches**

You can now use an arbitrary [OCI](https://opencontainers.org) registry as a build
cache:

```console
$ spack mirror add my_registry oci://user/image # Dockerhub
$ spack mirror add my_registry oci://ghcr.io/haampie/spack-test # GHCR
$ spack mirror set --push --oci-username ... --oci-password ... my_registry  # set login credentials
$ spack buildcache push my_registry [specs...]
```

> And you can add optionally add a base image to get *runnable* images:

Author: ```suggestion
And you can optionally add a base image to get *runnable* images:
```

   your favorite args to `spack find` as `spack f`:

   ```yaml
   config:
     aliases:
       f: find -lv
   ```

   See #17229.

10. **Improved autoloading of modules**

    Spack 0.20 was the first release to enable autoloading of direct dependencies in
    module files.

    The downside of this was that `module avail` and `module load` tab completion would
    show users too many modules to choose from, and many users disabled generating
    modules for dependencies through `exclude_implicits: true`. Further, it was
    necessary to keep hashes in module names to avoid file name clashes.

    In this release, you can start using `hide_implicits: true` instead, which exposes
    only explicitly installed packages to the user, while still autoloading
    dependencies. On top of that, you can safely use `hash_length: 0`, as this config
    now only applies to the modules exposed to the user -- you don't have to worry about
    file name clashes for hidden dependencies.

   Note: for `tcl` this feature requires Modules 4.7 or higher

11. **Updated container labeling**

    Nightly Docker images from the `develop` branch will now be tagged as `:develop` and

>     `:nightly` The `:latest` tag no longer be associated with `:develop`, but with the

Author: ```suggestion
    `:nightly`. The `:latest` tag is no longer associated with `:develop`, but with the
```

    latest stable release. Releases will be tagged with `:{major}`, `:{major}.{minor}`

>     and `:{major}.{minor}.{patch}` `ubuntu:18.04` has also been removed from the list of

Author: ```suggestion
    and `:{major}.{minor}.{patch}`. `ubuntu:18.04` has also been removed from the list of
```
